libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -582,9 +645,702 @@ class to_plugin2 final : public virtual operator_factory_plugin {
+  }
+
+private:
+  auto get() -> caf::result<table_slice> {
+    if (puts_.empty()) {
+      auto rp = self_->make_response_promise<table_slice>();
+      gets_.emplace_back(rp);
+      check_termination();
+      return rp;
+    }
+    auto slice = std::move(puts_.front().first);
+    puts_.front().second.deliver();
+    puts_.pop_front();
+    return slice;
+  }
+
+  auto put(table_slice slice) -> caf::result<void> {
+    if (gets_.empty()) {
+      auto rp = self_->make_response_promise<void>();
+      puts_.emplace_back(std::move(slice), rp);
+      return rp;
+    }
+    gets_.front().deliver(std::move(slice));
+    gets_.pop_front();
+    return {};
+  }
+
+  void query_files() {
+    add_actor_callback(
+      self_, fs_->GetFileInfoAsync(std::vector{root_path_}),
+      [this](arrow::Result<std::vector<arrow::fs::FileInfo>> infos) {
+        if (not infos.ok()) {
+          diagnostic::error("{}", infos.status().ToStringWithoutContextLines())
+            .primary(args_.url)
+            .emit(*dh_);
+          return;
+        }
+        TENZIR_ASSERT(infos->size() == 1);
+        auto root_info = std::move((*infos)[0]);
+        switch (root_info.type()) {
+          case arrow::fs::FileType::NotFound:
+            // TODO: Do we want to allow this? Maybe if `watch=true`?
+            diagnostic::error("`{}` does not exist", root_path_)
+              .primary(args_.url)
+              .emit(*dh_);
+            return;
+          case arrow::fs::FileType::Unknown:
+            diagnostic::error("`{}` is unknown", root_path_)
+              .primary(args_.url)
+              .emit(*dh_);
+            return;
+          case arrow::fs::FileType::File:
+            // TODO: Do we want to warn when the glob continues here?
+            process_file(std::move(root_info));
+            got_all_files();
+            return;
+          case arrow::fs::FileType::Directory:
+            auto sel = arrow::fs::FileSelector{};
+            sel.base_dir = root_path_;
+            sel.recursive = true;
+            auto gen = fs_->GetFileInfoGenerator(sel);
+            iterate_files(
+              self_, std::move(gen),
+              [this](arrow::Result<arrow::fs::FileInfoVector> files) {
+                if (not files.ok()) {
+                  diagnostic::error(
+                    "{}", files.status().ToStringWithoutContextLines())
+                    .primary(args_.url)
+                    .emit(*dh_);
+                  return;
+                }
+                if (files->empty()) {
+                  got_all_files();
+                  return;
+                }
+                for (auto& file : *files) {
+                  process_file(std::move(file));
+                }
+              });
+            return;
+        }
+        TENZIR_UNREACHABLE();
+      });
+  }
+
+  void process_file(arrow::fs::FileInfo file) {
+    // TODO: What if the glob matches a directory? Should the directory be
+    // processed in it's entirety?
```

Comment:
I think we should warn about this.

Reply:
I did not come up with a clear picture on when we would want to warn here an when not, because globs that are meant for files (and will also lead to processing files) often also match directories. So I say we just don't do anything about this for now.

(And just to recap: We decided to not process directories in their entirety just because they match the glob)